packages/database/src/server/changeHandlers/entityHierarchyCacher/EntityParentChildRelationBuilder.js

```diff
@@ -0,0 +1,226 @@
+   * @param {*} hierarchyId
+   * @param {*} parentIds
+   * @param {*} childrenAlreadyCached
+   * @returns
+   */
+  async generateViaCanonical(hierarchyId, parentIds, childrenAlreadyCached = new Set()) {
+    const canonicalTypes = await this.getCanonicalTypes(hierarchyId);
+    const entities = await this.models.entity.find({
+      parent_id: parentIds,
+      type: canonicalTypes,
+    });
+    const entityParentChildRelations = entities
+      .map(e => ({
+        parent_id: e.parent_id,
+        child_id: e.id,
+        entity_hierarchy_id: hierarchyId,
+      }))
+      // If the relation is already generated (could be that it is already a child of another parent)
+      // it should be ignored
+      .filter(relation => !childrenAlreadyCached.has(relation.child_id));
+
+    if (entityParentChildRelations.length === 0) {
+      return [];
+    }
+
+    return this.insertRelations(entityParentChildRelations);
+  }
+
+  async insertRelations(entityParentChildRelations) {
+    // If the relation is already generated, it should be ignored
+    await this.models.entityParentChildRelation.createMany(entityParentChildRelations, {
+      onConflictIgnore: ['entity_hierarchy_id', 'parent_id', 'child_id'],
+    });
+
+    return entityParentChildRelations.map(e => [e.parent_id, e.child_id]);
+  }
+
+  /**
+   * @private
+   */
+  async countEntityRelationChildren(hierarchyId, parentIds) {
+    return this.models.entityRelation.count({
+      parent_id: parentIds,
+      entity_hierarchy_id: hierarchyId,
+    });
+  }
+
+  async getCanonicalTypes(hierarchyId) {
+    const entityHierarchy = await this.models.entityHierarchy.findById(hierarchyId);
+    const { canonical_types: customCanonicalTypes } = entityHierarchy;
+    const canonicalTypes =
+      customCanonicalTypes && customCanonicalTypes.length > 0
+        ? customCanonicalTypes
+        : Object.values(ORG_UNIT_ENTITY_TYPES);
+
+    return canonicalTypes;
+  }
+
+  /**
+   * Delete the obsolete relations for a level of the hierarchy
+   * @param {*} hierarchyId
+   * @param {*} parentIds
+   * @param {*} validParentChildIdPairs
+   */
+  async deleteObsoleteRelationsForParents(hierarchyId, parentIds, validParentChildIdPairs) {
+    const valuesList = validParentChildIdPairs.map(() => '(?, ?)').join(', ');
+    const values = validParentChildIdPairs.flatMap(pair => pair);
+
+    // Delete any relations that:
+    // - Belong to the parentIds
+    // - Are not in the validParentChildIdPairs - which are the latest valid relations for this level
+    await this.models.database.executeSql(
+      `
+      DELETE FROM entity_parent_child_relation 
+      WHERE entity_hierarchy_id = ? 
+        AND parent_id IN (${parentIds.map(() => '?').join(', ')})
+        AND (parent_id, child_id) NOT IN (
+          SELECT * FROM (VALUES ${valuesList}) AS pairs(parent_id, child_id)
+        )
+      RETURNING parent_id, child_id
+    `,
+      [hierarchyId, ...parentIds, ...values],
```

Comment:
```suggestion
      [hierarchyId],
```
I don't think we're using those right?

Reply:
We do use them actually